src/raftstore/store/read_queue.rs

```diff
@@ -0,0 +1,260 @@
+        ReadIndexRequest {
+            id,
+            cmds,
+            renew_lease_time,
+            read_index: None,
+        }
+    }
+}
+
+impl Drop for ReadIndexRequest {
+    fn drop(&mut self) {
+        let dur = (monotonic_raw_now() - self.renew_lease_time)
+            .to_std()
+            .unwrap();
+        RAFT_READ_INDEX_PENDING_DURATION.observe(duration_to_sec(dur));
+    }
+}
+
+#[derive(Default)]
+pub struct ReadIndexQueue {
+    reads: VecDeque<ReadIndexRequest>,
+    ready_cnt: usize,
+    // How many requests are handled.
+    handled_cnt: usize,
+    // map[uuid] -> offset in `reads`.
+    contexts: HashMap<Uuid, usize>,
+}
+
+impl ReadIndexQueue {
+    pub fn notify_all_removed(&mut self, region_id: u64) {
+        let mut removed = 0;
+        for mut read in self.reads.drain(..) {
+            removed += read.cmds.len();
+            for (_, cb) in read.cmds.drain(..) {
+                apply::notify_req_region_removed(region_id, cb);
+            }
+        }
+        RAFT_READ_INDEX_PENDING_COUNT.sub(removed as i64);
+        self.contexts.clear();
+        self.ready_cnt = 0;
+        self.handled_cnt = 0;
+    }
+
+    pub fn clear_uncommitted(&mut self, term: u64) {
+        let mut removed = 0;
+        for mut read in self.reads.drain(self.ready_cnt..) {
+            self.contexts.remove(&read.id);
+            removed += read.cmds.len();
+            for (_, cb) in read.cmds.drain(..) {
+                apply::notify_stale_req(term, cb);
+            }
+        }
+        RAFT_READ_INDEX_PENDING_COUNT.sub(removed as i64);
+    }
+
+    pub fn push_back(&mut self, read: ReadIndexRequest, is_leader: bool) {
+        if !is_leader {
+            let offset = self.handled_cnt + self.reads.len();
+            self.contexts.insert(read.id, offset);
+        }
+        self.reads.push_back(read);
+    }
+
+    pub fn back_mut(&mut self) -> Option<&mut ReadIndexRequest> {
+        self.reads.back_mut()
+    }
+
+    pub fn advance_leader_reads<T>(&mut self, states: T) -> Option<Timespec>
+    where
+        T: IntoIterator<Item = (Uuid, u64)>,
+    {
+        let mut ts = None;
+        for (uuid, index) in states {
+            assert_eq!(uuid, self.reads[self.ready_cnt].id);
+            self.reads[self.ready_cnt].read_index = Some(index);
+            ts = Some(self.reads[self.ready_cnt].renew_lease_time);
+            self.ready_cnt += 1;
+        }
+        ts
+    }
+
+    /// update the read index of the requests that before the specified id.
+    pub fn advance_replica_reads<T>(&mut self, states: T)
+    where
+        T: IntoIterator<Item = (Uuid, u64)>,
+    {
+        let (mut min_changed_offset, mut max_changed_offset) = (usize::MAX, 0);
+        for (uuid, index) in states {
+            if let Some(offset) = self.contexts.remove(&uuid) {
+                let offset = offset.checked_sub(self.handled_cnt).unwrap();
+                self.ready_cnt = cmp::max(self.ready_cnt, offset + 1);
```

Comment:
I think it can be calculated later by `cmp::max(self.ready_cnt, max_changed_offset + 1)`.